trestle/core/remote/cache.py

```diff
@@ -0,0 +1,487 @@
+
+    def _sync_cache(self) -> None:
+        auth = None
+        if self._username is not None and self._password is not None:
+            auth = HTTPBasicAuth(self._username, self._password)
+        request = requests.get(self._furl.url, auth=auth)
+        if request.status_code == 200:
+            result = request.json()
+            if result is None:
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), str(self._inst_cache_path))
+            if result['isBinary']:
+                raise NotImplementedError('Binary files are not supported!')
+            else:
+                self._inst_cache_path.write_text(result['text'])
+        else:
+            raise TrestleError(f'Query failed to run by returning code of ' f'{request.status_code}. {self._query}')
+
+
+class SFTPFetcher(FetcherBase):
+    """Fetcher for SFTP content."""
+
+    # STFP method: https://stackoverflow.com/questions/7563496/open-a-remote-file-using-paramiko-in-python-slow#7563551
+    # For SFTP fetch into memory.
+    def __init__(
+        self,
+        trestle_root: pathlib.Path,
+        uri: str,
+        settings: Settings,
+        refresh: bool = False,
+        fail_hard: bool = False,
+        cache_only: bool = False
+    ) -> None:
+        """Initialize STFP fetcher."""
+        super().__init__(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+        # Is this a valid uri, however? Username and password are optional, of course.
+        u = parse.urlparse(self._uri)
+        if not u.hostname:
+            logger.error(f'Malformed URI, cannot parse hostname in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing hostname {self._uri}')
+        if not u.path:
+            logger.error(f'Malformed URI, cannot parse path in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing file path {self._uri}')
+        if u.password and not u.username:
+            logger.error(f'Malformed URI, password found but username missing in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: password found but username missing {self._uri}')
+
+        localhost_cached_dir = self._trestle_cache_path / u.hostname
+        # Skip any number of back- or forward slashes preceding the url path (u.path)
+        path_parent = pathlib.Path(u.path[re.search('[^/\\\\]', u.path).span()[0]:]).parent
+        localhost_cached_dir = localhost_cached_dir / path_parent
+        localhost_cached_dir.mkdir(parents=True, exist_ok=True)
+        self._inst_cache_path = localhost_cached_dir / pathlib.Path(pathlib.Path(u.path).name)
+
+    def _sync_cache(self) -> None:
+        u = parse.urlparse(self._uri)
+        client = paramiko.SSHClient()
+
+        if 'SSH_KEY' in os.environ and self._refresh:
+            ssh_key_file = os.environ['SSH_KEY']
+            try:
+                client.load_host_keys(ssh_key_file)
```

Comment:
I don't think this is doing what you think it is. paramiko should automatically load ssh-keys from standard locations. I think this is adding to (in memory) the known_hosts file. 

There is a connect parameter called Pkey which I think does what you are intending to do. Have you confirmed this behaviour yet (keeping in mind what paramiko will load from an .ssh directory?).

Reply:
Oh, yes, forgot I was going to fix that weeks ago. That dawned on me shortly after I pushed it and gave up for the day, then promptly forgot about it. This is indeed about the user's known_hosts file, not the private key. I'll fix that.

Paramiko automatically uses the agent if it's up and if the right key is active there. I didn't get to loading a non-standard key file in a non-standard location, and I don't know if paramiko will force a password prompt if it does find the right key in ~/.ssh but it hasn't been unlocked in the agent. I'll get onto this.

Reply:
Fixed! Note we only have the usual paths to look for the host keys, i.e., ~/.ssh/known_hosts for Mac/Linux. We have no facility to supply a custom path to consult.